doc/html2xhtml.1
.TH html2xhtml 1 ""
.SH NAME
html2xhtml \- Converts HTML files to XHTML
.SH SYNTAX 

.B html2xhtml
[
.I filename
] [
.I options
] 

.SH DESCRIPTION

Html2xhtml is a command-line tool that converts HTML files to
XHTML files. The path of the HTML input file can be provided
as a command-line argument. If not, it is read from stdin.
.PP
Xhtml2xhtml tries always to generate valid XHTML files.
It is able to correct many common errors in input HTML
files without loose of information. However, for some errors, html2xhtml
may decide to loose some information in order to generate a valid XHTML output.
This can be avoided with the 
.I -e 
option, which allows html2xhtml to
generate non-valid output in these cases.
.PP
Html2xhtml can generate the XHTML output compliant to one of
the following document types:
XHTML 1.0 (Transitional, Strict and Frameset), XHTML 1.1,
XHTML Basic and XHTML Mobile Profile. 
.SH OPTIONS
.if n .ta 2.8i
.if t .ta 2.1i

The command line options/arguments are:
.IP \fBfilename\fR 20
Read the HTML input from
.I filename
(optional argument). If this argument is not provided, 
the HTML input is read from standard input.
.IP \fB-o\ \fIfilename\fR 20
Output XHTML file. The file is overwritten if it
exists. If not provided, the output is written to
standard output.
.IP \fB-e\fR 20
Instructs the program to propagate input chunks to 
the output even if it is unable to adapt them to
the output XHTML doctype. Using this option, 
the XHTML output may be non-valid. Not using this
option, some input data could be removed from the
output in some [rare] cases.
.IP \fB-t\ \fIoutput-doctype\fR 20
Doctype of the output XHTML file. If not specified,
the program selects automatically either 
XHTML 1.0 Transitional or XHTML 1.0 Frameset 
depending on the input. Current available
doctypes are:
.br
 o \fItransitional\fR
XHTML 1.0 Transitional
.br
 o \fIframeset\fR
XHTML 1.0 Frameset
.br 
 o \fIstrict\fR
XHTML 1.0 Strict
.br 
 o \fI1.1\fR
XHTML 1.1
.br 
 o \fIbasic-1.0\fR
XHTML Basic 1.0
.br 
 o \fIbasic-1.1\fR
XHTML Basic 1.1
.br 
 o \fImp\fR
XHTML Mobile Profile
 o \fIprint-1.0\fR
XHTML Print 1.0
.br 
.IP \fB--ics\ \fIinput_charset\fR 20
Character set of the input document. This option overrides the
default input character set detection mechanism.
.IP \fB--ocs\ \fIoutput_charset\fR 20
Character set for the output XHTML document. If this option is not
present, the character set of the input is used as default.
.IP \fB--lcs\fR 20
Dump the list of available character set aliases and exit html2xhtml.
No conversion is performed when this option is present.
.IP \fB-l\ \fIline_length\fR 20
Number of characters per line. The default value is 80.
It must be greater or equal to 40, otherwise the 
parameter is ignored.
.IP \fB-b\ \fItab_length\fR 20
Tab length in number of characters. It must be a number
between 0 and 16, otherwise the parameter is ignored.
Use 0 to avoid indentation in the output.
.IP \fB--preserve-space-comments\fR 20
Use this option to preserve white spaces, tabs and 
ends of lines in HTML comments. The default, if not 
provided, is to rearrange spacing.
.IP \fB--no-protect-cdata\fR 20
Enclose CDATA sections in "script" and "style" following
the XHTML 1.0 specification (using "<!CDATA[[" and
"]]>"). It might be incompatible with some browsers.
The default in this version is to enclose CDATA sections
using "//<!CDATA[[" and "//]]>", because major browsers
handle it properly. 
.IP \fB--compact-block-elements\fR 20
No white spaces or line breaks are written between  
the start tag of a block element and the start tag 
of its first enclosed inline element (or character 
data) and between the end tag of its last enclosed 
inline element (or character data) and the end tag 
of the block element. By default, if this option is
not set, a new line character and indentation is
written between them.
.IP \fB--compact-empty-elm-tags\fR 20
Do not write a whitespace before the slash for empty element
tags (i.e. write "<br/>" instead of the default "<br />").
Note that although both notations are correct in XML,
the XHTML 1.0 standard recommends the latter to improve
compatibility with old browsers.
.IP \fB--empty-elm-tags-always\fR 20
By default, empty element tags are written only for
elements declared as empty in the DTD. This option
makes any element not having content to be written
with the empty element tag, even if it is not declared
as empty in the DTD. This option may cause problems
when the XHTML document is opened by browsers in
HTML (tag soup) mode.
.IP \fB--dos-eol\fR 20
Write the output XHTML file with DOS--style (CRLF) end of line,
instead of the default UNIX--style end of line.
Both end of line styles are allowed by the XML recommendation.
.IP \fB--generate-snippet\fR 20
Treat the input as an HTML fragment instead of
a full document.
The output will also be a snippet and will not contain either the XML
and doctype declarations or the html, head and body elements.
.IP \fB--system-dtd-prefix\ \fIprefix\fR 20
Use a SYSTEM doctype declaration instead of the PUBLIC default one.
The system location is created by concatenating the provided prefix
and the name of the DTD file.
Prefix values can be relative local paths or URIs.
.IP \fB--help\fR 20
Show a brief help message and exit.
.IP \fB--version\fR 20
Show the version number and exit.

.SH NOTE ON CHARACTER SETS

Since version 1.1.2, html2xhtml is able to parse and write
HTML and XHTML documents using the most popular character sets / encodings.
It is also able to read the input document using a given
character set and generate an output that uses a different
character set. The iconv implementation in the GNU C library
is used with that purpose.
.PP
Any IANA-registered character set that is supported by the
iconv library may be used. When naming a character set, any
IANA--approved alias for it may be used. The full list of
aliases recognised by html2xhtml can be obtained with the
\fB--lcs\fR command-line option.
.PP
If the character set of the input document is not specified,
html2xhtml tries to guess it automatically.
If the character set of the output document is not specified,
html2xhtml writes the output using the same character set
as the input document.

.SH NOTE ON END OF LINE CHARACTES

By default, the UNIX-style one-byte end of line is used. It can
be changed to DOS-style CRLF end of line with the
\fB--dos-eol\fR command-line option.
.PP
However, when the program is compiled in the MinGW environment
and the output is sent to standard output, the output is automatically
converted by the environment to CRLF by default. Do not use the
\fB--dos-eol\fR command-line option in that situation.
When the output is sent to a file with the \fB-o\fR command-line
option, the output is as expected (UNIX-style by default), and
the \fB--dos-eol\fR option may be used.

.SH ACKNOWLEDGMENTS
.na 
.nf

Program developer up to current version:
Jesus Arias Fisteus <jaf@it.uc3m.es>
.PP
The first working version of this program has been developed as 
a Master Thesis at the University of Vigo (Spain) [http://www.uvigo.es],
advised by:
.PP
Rebeca Diaz Redondo
Ana Fernandez Vilas
.PP
Copyright 2000-2001 by Jesus Arias Fisteus, Rebeca Diaz Redondo, Ana
Fernandez Vilas.
Copyright 2002-2009 by Jesus Arias Fisteus





ChangeLog
2025-06-02  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml Released version 1.4
	* html2xhtml New command line option to generate
	a SYSTEM DTD declaration.
	* html2xhtml Several bugfixes related to memory management.

2015-09-23  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml Released version 1.3
	* html2xhtml Improved solution to the vsnprintf problem
	* html2xhtml New snippet mode.
	* html2xhtml Web API example clients for BASH, Node.js and Python.
	* html2xhtml Fixed a segfault error.
	* html2xhtml Updated the autotools files to recent versions.

2010-04-07  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml Released version 1.1.2-2
	* html2xhtml Bugfix related to the use of vsnprintf

2009-11-05  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml Released as stable, version 1.1.2

2009-10-11  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml The DTD coder module implements now a cache of
	downloaded DTD entities. The User-Agent header of that feature
	has also been changed to avoid being blocked by W3C servers.

2009-07-03  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml More robust HTML parser, using the Bison's error
	recovery mechanism.
	* html2xhtml Released as 1.1.2-rc3

2009-07-02  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml More robust handling of invalid chars

2009-06-30  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml Several bugfixes.
	* html2xhtml Released as 1.1.2-rc2

2009-05-28  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml Updated DTD data.

2009-04-13  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml Released as 1.1.2-rc1

2009-04-08  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml End-of-line (UNIX-like or DOS-like) characters can be
	chosen with a command-line parameter.
	* html2xhtml Improved documentation.

2009-03-26  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml Updated DTD data.

2009-03-25  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml Redesigned CGI-handling code. Support to both
	multipart/form-data input and direct (text/html) input.

2009-03-10  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml New --compact-empty-elem-tags option. Renamed
	option --empty-elm-tags-always.
	* html2xhtml Support for many input/output character sets using iconv.

2009-03-03  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml Adapted the lexer to recent flex versions (2.5.35)

2008-07-28  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml New test system, to help to detect regressions.

2008-07-27  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml Empty element tags written only for elements 
	declared as empty in DTD.

2008-06-26  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml Updated DTD data.
	* html2xhtml Ported dtd_query from the old source tree.

2008-06-17  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml Released as stable, version 1.1.1.

2008-05-29  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml Removed the REJECT rule from the parser.

2008-05-22  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml Fixed a bug in the CGI related to charset.

2008-04-17  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml Released as 1.1b6

2008-04-08  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml Fixed two bugs related to incorrect memory access.

2008-03-31  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml Fixed a bug when "<" and "&" follow "]" in CDATA.
	* html2xhtml Fixed whitespace problems with entity references.
	* html2xhtml Entity reference "percnt" is replaced by "%". 
	* html2xhtml Flexibility added to the parser: allows closing 
	XML PIs with ">" instead of "%>"
	
2008-03-05  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml Released as 1.1b5

2008-03-03  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml Fixed a bug when an end tags appears inside
	a script. The end tag was not properly wrapped inside a CDATA
	section.
	* html2xhtml Fixed a bug in the CGI: comments were not properly
	escaped.
	* html2xhtml Changed the result page of the CGI to make it
	look closer to the other pages of the html2xhtml website.

2008-01-16  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml Released as 1.1b4

	* html2xhtml Fixed some bugs with Windows carriage returns.

	* html2xhtml Fixed a bug when line breaks appeared inside
	attribute values.

	* html2xhtml Added the --compact-block-elements option.

2008-01-04  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml New output formatting code. Now indentation 
	length and line length are configurable. Line wrapping is 
	now better.

	* html2xhtml Integrated the code of the CGI version of the
	program in the executable file html2xhtml. Define the
	macro WITH_CGI to activate it at compilation time. The
	executable will decide at run time if it works in command
	line or as a CGI.

	* html2xhtml Better support of XHTML input.

	* html2xhtml Better support for CDATA sections.
	
	* html2xhtml Fixed several other minor issues.

2007-10-30  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml Fixed a bug regarding whitespaces: spacing was
	removed in situations like "<b>1</b> <b>2</b>"

2007-09-16  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml Added support for XHTML Print 1.0

	* html2xhtml Bug fixes and help message improved

	* html2xhtml Now converts from XHTML to XHTML

	* html2xhtml Released as 1.1b3
	
2007-09-15  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml Added support for XHTML Basic 1.1 and updated DTDs

	* html2xhtml Released as 1.1b2
	
2007-05-19  Jesus Arias Fisteus <jaf@it.uc3m.es>

	* html2xhtml First release under the GNU GPL license (1.1b1)
	
	* html2xhtml First version with autotools-based build system
	
	* html2xhtml Removed from the source tree the CGI and DLL
	versions, dtd_test and xhtmlpedia.

	
	



src/html2xhtml.c
/***************************************************************************
 *   Copyright (C) 2007 by Jesus Arias Fisteus                             *
 *   jaf@it.uc3m.es                                                        *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/


/*
 * html2xhtml.c
 *
 * 'main' module for the command line version of html2xhtml.
 *
 *
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <stdio.h>
#include <string.h>

#include "procesador.h"
#include "dtd_names.h"
#include "dtd_util.h"
#include "dtd.h"
#include "mensajes.h"
#include "tree.h"
#include "xchar.h"
#include "params.h"
#include "charset.h"

#ifdef WITH_CGI
#include "cgi.h"
#endif

/* parser Yacc / lex */
int yyparse(void);
void parser_set_input(FILE *input);  /* in html.l */

void print_version(void);

static void process_parameters(int argc,char **argv);
static void help(void);
static void print_doctypes(void);
static void print_doctype_keys(void);

int main(int argc,char **argv)
{
  size_t preload_read;
  int yyparse_result;

#ifdef WITH_CGI
  const char *preload_buffer;
#endif

  tree_init();

#ifdef WITH_CGI
  cgi_check_request();
  if (cgi_status < 0) {
    cgi_write_error_bad_req();
    return 0;
  }
#endif

  params_set_defaults();

#ifdef WITH_CGI
  if (!cgi_status)
    process_parameters(argc, argv);

  preload_buffer = charset_init_preload(param_inputf, &preload_read);

  if (cgi_status > 0)
    cgi_process_parameters(&preload_buffer, &preload_read);

  charset_auto_detect(preload_read);
  charset_preload_to_input(param_charset_in, preload_read);
  if (cgi_status == CGI_ST_MULTIPART)
    charset_cgi_boundary(boundary, boundary_len);
#else
  /* process command line arguments */
  process_parameters(argc, argv); 
  charset_init_preload(param_inputf, &preload_read);
  charset_auto_detect(preload_read);
  charset_preload_to_input(param_charset_in, preload_read);
#endif

  /* intialize the converter */
  saxStartDocument();
  if (param_inputf != stdin)
    parser_set_input(param_inputf);

  /* parse the input file and convert it */
  yyparse_result = yyparse();

  if (yyparse_result) {
    EXIT("Unrecoverable parse error");
  }

  charset_close();
  saxEndDocument();

#ifdef WITH_CGI
  if (!cgi_status) {
    /* write the output */
    if (writeOutput()) 
      EXIT("Bad state in writeOutput()");

    /* close de output file */
    if (param_outputf != stdout)
      fclose(param_outputf);
  } else {
    cgi_write_output();
  }
#else
  /* write the output */
  if (writeOutput()) 
    EXIT("Bad state in writeOutput()");
  
  /* close de output file */
  if (param_outputf != stdout)
    fclose(param_outputf);
#endif

  /* show final messages */
  write_end_messages();
  freeMemory();

  return 0;
}

static void process_parameters(int argc, char **argv)
{
  int i, fich, tmpnum;

  /* process command line arguments */
  for (i=1, fich=0; i<argc; i++) {
    if (!strcmp(argv[i], "-e")) {
      param_strict = 0;
    } else if (!strcmp(argv[i], "-t") && ((i+1) < argc)) {
      param_doctype = dtd_get_dtd_index(argv[++i]);
    } else if (!strcmp(argv[i], "-o") && ((i+1) < argc)) {
      /* open the output file */
      param_outputf = fopen(argv[++i], "wb");
      if (!param_outputf) {
        perror("fopen");
        EXIT("Could not open the output file for writing");
      }
    } else if (!strcmp(argv[i], "-l") && ((i+1) < argc)) {
      tmpnum= atoi(argv[++i]);
      if (tmpnum >= 40)
        param_chars_per_line = tmpnum; 
    } else if (!strcmp(argv[i], "-b") && ((i+1) < argc)) {
      tmpnum= atoi(argv[++i]);
      if (tmpnum >= 0 && tmpnum <= 16)
        param_tab_len = tmpnum; 
    } else if (!strcmp(argv[i], "--preserve-space-comments")) {
      param_pre_comments = 1;
    } else if (!strcmp(argv[i], "--no-protect-cdata")) {
      param_protect_cdata = 0;
    } else if (!strcmp(argv[i], "--compact-block-elements")) {
      param_compact_block_elms = 1;
    } else if (!strcmp(argv[i], "--compact-empty-elem-tags")) {
      param_compact_empty_elm_tags = 1;
    } else if (!strcmp(argv[i], "--empty-elm-tags-always")) {
      param_empty_tags = 1;
    } else if (!strcmp(argv[i], "--dos-eol")) {
      param_crlf_eol = 1;
    } else if (!strcmp(argv[i], "--system-dtd-prefix") && ((i+1) < argc)) {
      param_system_dtd_prefix = argv[++i];
    } else if (!fich && argv[i][0]!='-') {
      fich = 1;
      param_inputf = fopen(argv[i], "r");
      if (!param_inputf) {
        perror("fopen");
        EXIT("Could not open the input file for reading");
      }
    } else if (!strcmp(argv[i], "--help") || !strcmp(argv[i], "-h")) {
      help();
      exit(0);
    } else if (!strcmp(argv[i], "--version") || !strcmp(argv[i], "-v")) {
      print_version();
      exit(0);
    } else if (!strcmp(argv[i], "-L")) {
      print_doctype_keys();
      exit(0);
    } else if (!strcmp(argv[i], "--ics") && ((i+1) < argc)) {
      param_charset_in = charset_lookup_alias(argv[++i]);
      if (!param_charset_in) {
        EPRINTF1("Trying to set input character set: %s\n", argv[i]);
        EXIT("Unsupported character set");
      }
    } else if (!strcmp(argv[i], "--ocs") && ((i+1) < argc)) {
      param_charset_out = charset_lookup_alias(argv[++i]);
      if (!param_charset_out) {
        EPRINTF1("Trying to set output character set: %s\n", argv[i]);
        EXIT("Unsupported character set");
      }
    } else if (!strcmp(argv[i], "--lcs")) {
      charset_dump_aliases(stdout);
      exit(0);
    } else if (!strcmp(argv[i], "--generate-snippet")) {
      param_generate_snippet = 1;
    } else {
      help();
      exit(1);
    }
  } 
}

int yyerror(char *e)
{
  /* Let bison's error recovery mechanisms work */
  WARNING(e);
  return 0;
}

void exit_on_error(char *msg)
{
#ifdef WITH_CGI
    /* this function exits the program */
  if (cgi_status) {  
    cgi_write_error(msg);
  } else { 
#endif

#ifdef MSG_DEBUG
    fprintf(stderr,"!!%s(%d)[l%d]: %s\n",__FILE__,__LINE__,
            parser_num_linea,msg);
#else
    fprintf(stderr,"Error [line %d]: %s\n", parser_num_linea, msg);    
#endif
    write_end_messages();

#ifdef WITH_CGI
  }
#endif

  freeMemory();
  exit(1);
}

static void help(void)
{
  print_version();
  fprintf(stderr, "Usage:\n");
  fprintf(stderr, "html2xhtml [<input_html_file>] [-t <output_doctype_key>] [-e]\n");
  fprintf(stderr, "           [-o <output_file>]\n");
  fprintf(stderr, "           [--ics <input_charset>] [--ocs <output_charset>]\n");
  fprintf(stderr, "           [-l <line_length>] [-b <tab_length>]\n");
  fprintf(stderr, "           [--preserve-space-comments] [--no-protect-cdata]\n");
  fprintf(stderr, "           [--compact-block-elements] [--empty-elm-tags-always]\n");
  fprintf(stderr, "           [--compact-empty-elm-tags] [--dos-eol]\n");
  fprintf(stderr, "           [--generate-snippet]\n");
  fprintf(stderr, "           [--system-dtd-prefix <dtd_location_prefix>]\n");
  fprintf(stderr, "\n");
  print_doctypes();
}

static void print_doctypes(void)
{
  int i;
  fprintf(stderr, "Setting output doctype:\n");
  for (i = 0; i < XHTML_NUM_DTDS; i++) {
    fprintf(stderr, "'-t %s': document type %s.\n",
            dtd_key[i], dtd_name[i]);
  }
}

static void print_doctype_keys(void)
{
  int i;
  for (i = 0; i < XHTML_NUM_DTDS; i++) {
    fprintf(stdout, "%s\n", dtd_key[i]);
  }
}

void print_version(void)
{
#ifndef WITH_CGI
  fprintf(stderr, "html2xhtml version %s\n", VERSION);
#else
  fprintf(stderr, "html2xhtml version %s [cgi-enabled]\n", VERSION);
#endif
  fprintf(stderr, "DTD data based on DTDs as available on %s\n\n",
          DTD_SNAPSHOT_DATE);
}



